utils/in_mem_item_tree/src/primitives/detail.rs

```diff
@@ -0,0 +1,262 @@
+
+use super::{arena::Arena, node_id::NodeId, DataItem, Error, Flavor};
+
+/// The trait that is referenced by the `Flavor` trait. `Arena` holds an instance of such
+/// an observer and notifies it about the corresponding modifications.
+pub trait ModificationObserver<T: DataItem>: Clone + std::fmt::Debug + Default {
+    fn on_node_added(&mut self, node_id: NodeId, item_id: <T as DataItem>::Id)
+        -> Result<(), Error>;
+    fn on_node_removed(&mut self, item_id: &<T as DataItem>::Id);
+    fn on_whole_arena_cleared(&mut self);
+}
+
+/// `Arena` uses this trait to get access to the item-id-to-node-id map that is held within
+///  the observer, if any.
+pub trait ItemIdMapHolder<T: DataItem> {
+    fn item_id_map(&self) -> &BTreeMap<<T as DataItem>::Id, NodeId>;
+}
+
+/// This is the observer that always holds the item-id-to-node-id map, which allows `Arena`
+/// to implement some extra methods.
+pub struct TrackingModificationObserver<T: DataItem>(BTreeMap<<T as DataItem>::Id, NodeId>);
+
+impl<T: DataItem> ModificationObserver<T> for TrackingModificationObserver<T> {
+    fn on_node_added(
+        &mut self,
+        node_id: NodeId,
+        item_id: <T as DataItem>::Id,
+    ) -> Result<(), Error> {
+        use std::collections::btree_map::Entry;
+
+        match self.0.entry(item_id) {
+            Entry::Vacant(e) => {
+                e.insert(node_id);
+                Ok(())
+            }
+            Entry::Occupied(_) => Err(Error::ItemAlreadyInArena {
+                item_id: item_id.to_string(),
+            }),
+        }
+    }
+
+    fn on_node_removed(&mut self, item_id: &<T as DataItem>::Id) {
+        let removed = self.0.remove(item_id).is_some();
+        debug_assert_or_log!(removed, "item id {item_id} wasn't in the map");
+    }
+
+    fn on_whole_arena_cleared(&mut self) {
+        self.0.clear();
+    }
+}
+
+// Implement Default/Debug/Clone by hand, because `derive` would automatically add the corresponding
+// trait bounds for T, which we don't need.
+impl<T: DataItem> Default for TrackingModificationObserver<T> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+impl<T: DataItem> std::fmt::Debug for TrackingModificationObserver<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("TrackingModificationObserver").field(&self.0).finish()
+    }
+}
+impl<T: DataItem> Clone for TrackingModificationObserver<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl<T: DataItem> ItemIdMapHolder<T> for TrackingModificationObserver<T> {
+    fn item_id_map(&self) -> &BTreeMap<<T as DataItem>::Id, NodeId> {
+        &self.0
+    }
+}
+
+/// This is the observer that only holds the item-id-to-node-id map in debug builds, to perform
+/// the corresponding runtime checks.
+pub struct DebugOnlyTrackingModificationObserver<T: DataItem> {
+    #[cfg(debug_assertions)]
```

Comment:
Just note that the overuse of cfg like this complicates testing. I wouldn't recommend doing this. I would prefer to have a run-time configurable flag. Up to you.

Reply:
> Just note that the overuse of cfg like this complicates testing. I wouldn't recommend doing this. I would prefer to have a run-time configurable flag. Up to you.

Since you don't insist, I'll leave it as is. IMO it won't complicate testing of the user code, because it basically adds some checks in debugs builds.